### PR TITLE
BLOC-483 -- Excluding tags

### PR DESCRIPTION
**Pull request by Andrei-Masilevich**

### CODE REVIEW

File: libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp [hunk 1]
```diff
@@ -433,6 +433,49 @@ class tags_api_impl
         return result;
     }
 
+    posts_crefs exclude(const posts_crefs& filtering_posts,
+                        const discussion_query& query,
+                        const std::function<bool(const tag_object&)>& tag_filter) const
+    {
+        auto rng_exclude = query.exclude_tags | boost::adaptors::transformed(utils::to_lower_copy)
```
Comment (Katkevich) [inline libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp]: Add `//` (comment) just after `auto rng_exclude = query.exclude_tags` to force clang-format to move each `| boost::adaptors` on separate line

File: libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp [hunk 2]
```diff
@@ -433,6 +433,49 @@ class tags_api_impl
...
+        };
+
+        std::set<post_cref, less_by_comment> exclude_set;
+        for (const auto& posts : exclude_posts_by_tags)
+            boost::copy(posts, std::inserter(exclude_set, end(exclude_set)));
+
+        posts_crefs result;
+        result.reserve(filtering_posts.size());
+
+        std::copy_if(begin(filtering_posts), end(filtering_posts), std::back_inserter(result),
```
Comment (Katkevich) [inline libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp]: Use boost::set_difference

File: libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp [hunk 3]
```diff
@@ -470,6 +517,11 @@ class tags_api_impl
         if (posts.empty())
             return result;
 
+        if (!query.exclude_tags.empty())
+            posts = exclude(posts, query, tag_filter);
```
Comment (Katkevich) [inline libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp]: You are already checking posts collection for emptyness inside `exclude` function so no need to check this twice. Moreover it is better to move this line `posts = exclude(posts, query, tag_filter);` before 
```
if (posts.empty())
    return result;
```

File: libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp [hunk 4]
```diff
@@ -433,6 +433,49 @@ class tags_api_impl
...
+        if (tags_exclude.empty())
+            return filtering_posts;
+
+        std::vector<posts_crefs> exclude_posts_by_tags;
+        exclude_posts_by_tags.reserve(tags_exclude.size());
+
+        boost::transform(tags_exclude, std::back_inserter(exclude_posts_by_tags),
+                         [&](const std::string& t) { return get_posts(t, tag_filter); });
+
+        if (exclude_posts_by_tags.empty())
```
Comment (Katkevich) [inline libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp]: No need to check `exclude_posts_by_tags` for emptyness

File: libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp (hunk 2 above)
Comment (Andrei-Masilevich) [reply libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp]: set_difference require that containers must be sorted.
But there is different sorting order for filtering_posts and exclude_set and almost no profit

File: libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp (hunk 4 above)
Comment (Andrei-Masilevich) [reply libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp]: To prevent  iterations in std::copy_if

**Pull request merged**